Libs/Core/ctkCoreSettings.cpp

```diff
@@ -0,0 +1,233 @@
+/*=========================================================================
+
+  Library:   CTK
+
+  Copyright (c) Kitware Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0.txt
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=========================================================================*/
+
+#include "ctkCoreSettings.h"
+
+// CTK includes
+#include <ctkPimpl.h>
+
+// Qt includes
+#include <QDebug>
+#include <QDir>
+#include <QFileInfo>
+#include <QVariant>
+
+
+//-----------------------------------------------------------------------------
+class ctkCoreSettingsPrivate
+{
+  Q_DECLARE_PUBLIC(ctkCoreSettings);
+protected:
+  ctkCoreSettings* const q_ptr;
+public:
+  ctkCoreSettingsPrivate(ctkCoreSettings& object);
+  virtual ~ctkCoreSettingsPrivate();
+
+  QString ApplicationHomeDirectory;
+  QString ApplicationHomePlaceholder;
+};
+
+//-----------------------------------------------------------------------------
+ctkCoreSettingsPrivate::ctkCoreSettingsPrivate(ctkCoreSettings& object)
+  : q_ptr(&object)
+{
+  this->ApplicationHomePlaceholder = "<APPLICATION_HOME_DIR>";
```

Comment:
I am wondering if we should also update the launcher to support expanding a custom placeholder.

By not doing so, I anticipate settings like `PYTHONPATHS` or `LibraryPaths` that are expected to be understood by the launcher will not be expanded properly after installing extensions.

Support for having launcher settings file with a section like this would do it I think:

```
[PlaceHolderAliases]
APPLAUNCHER_DIR=APPLICATION_HOME_DIR
```

Note: In the launcher documentation, placeholder are called "special string". See https://github.com/commontk/AppLauncher/blob/b1d6dad48b11c2770a72ad879812995213ccfc6e/Base/ctkAppLauncherSettings.h#L58-L69

Reply:
How would the custom placeholder get to a launcher settings ini file?
Does Slicer modify "c:\Users\andra\AppData\Local\NA-MIC\Slicer 4.11.0-2020-07-06\bin\SlicerLauncherSettings.ini" when you install extensions? Where?

Instead of introducing aliases, we could use the existing <APPLAUNCHER_DIR> placeholder in launcher settings ini files.

Reply:
When installing the extension, the Slicer-NNNN.ini file is modified (see [here](https://github.com/Slicer/Slicer/blob/6751d8e6d8cdc8e7c221ed2165d902cd0c066554/Base/QTCore/qSlicerExtensionsManagerModel.cxx#L629-L640)) to include the relevant settings so that the launcher Slicer, PythonSlicer and also in the Slicer [application](https://github.com/Slicer/Slicer/blob/6751d8e6d8cdc8e7c221ed2165d902cd0c066554/Base/QTCore/qSlicerCoreApplication.cxx#L231-L262) to support case of macOS can setup the environment.

Currently this paths are set here:

It is true that `<APPLAUNCHER_DIR>` can be used and that would nicely address what I discussed.

Reply:
See also https://github.com/Slicer/Slicer/blob/43300c7b9a3b27265c2bc736693cb881f05e3ea9/Base/QTCore/qSlicerCoreApplication.cxx#L373

Reply:
Is launcher setting supposed to be updated here?

https://github.com/Slicer/Slicer/blob/43300c7b9a3b27265c2bc736693cb881f05e3ea9/Base/QTCore/qSlicerExtensionsManagerModel.cxx#L507

But the only place I've found ExtensionsSettingsFilePath is set is here:

https://github.com/Slicer/Slicer/blob/43300c7b9a3b27265c2bc736693cb881f05e3ea9/Base/QTCore/qSlicerCoreApplication.cxx#L373

and it is set to q->slicerRevisionUserSettingsFilePath(), which is Slicer-NNN.ini, so for me it seems that SlicerLauncherSettings.ini is not updated. Is this true?

Does the launcher read Slicer-NNNN.ini to get additional paths?

Reply:
> so for me it seems that SlicerLauncherSettings.ini is not updated. Is this true?

This is correct

> Does the launcher read Slicer-NNNN.ini to get additional paths?

Yes

This happen thanks to this information read from the `SlicerLauncherSettings`:

```
[Application]
[...]
name=Slicer
revision=29205
organizationDomain=www.na-mic.org
organizationName=NA-MIC
```

the location of the `<appname>-<revision>` is inferred based on the `organizationDomain`, `organizationName` and `name` of key found in the `[Application]` group

Reply:
Thanks, I understand the mechanism then.

I've made the placeholder string configurable. This will allow us to use `<APPLAUNCHER_DIR>` as placeholder (special text), so the launcher will understand it, too.